src/include/units/ratio.h

```diff
@@ -86,64 +88,70 @@ struct ratio {
   }
 }
 
-template<std::intmax_t N>
-[[nodiscard]] constexpr ratio pow(const ratio& r)
-{
-  if constexpr(N == 0)
-    return ratio(1);
-  else if constexpr(N == 1)
-    return r;
-  else
-    return pow<N-1>(r) * r;
-}
-
 namespace detail {
 
-// sqrt_impl avoids overflow and recursion
-// from http://www.codecodex.com/wiki/Calculate_an_integer_square_root#C.2B.2B
-// if v >= place this will fail  (so we can't quite use the last bit)
-[[nodiscard]] constexpr std::intmax_t sqrt_impl(std::intmax_t v)
+[[nodiscard]] constexpr auto make_exp_align(const ratio& r, std::intmax_t alignment)
 {
-  // place = 0x4000 0000 for 32bit
-  // place = 0x4000 0000 0000 0000 for 64bit
-  std::intmax_t place = static_cast<std::intmax_t>(1) << (sizeof(std::intmax_t) * 8 - 2);
-  while (place > v) place /= 4;  // optimized by complier as place >>= 2
-
-  std::intmax_t root = 0;
-  while (place) {
-    if (v >= root + place) {
-      v -= root + place;
-      root += place * 2;
-    }
-    root /= 2;
-    place /= 4;
+  Expects(alignment > 0);
+  const std::intmax_t rem = r.exp % alignment;
+
+  if (rem == 0) {  // already aligned
+    return std::array{r.num, r.den, r.exp};
+  }
+
+  if (r.exp > 0) {  // remainder is positive
+    return std::array{r.num * ipow10(rem), r.den, r.exp - rem};
   }
-  return root;
+
+  // remainder is negative
+  return std::array{r.num, r.den * ipow10(-rem), r.exp - rem};
 }
 
-[[nodiscard]] constexpr auto make_exp_even(const ratio& r)
+template<std::intmax_t N>
+[[nodiscard]] constexpr ratio root(const ratio& r) requires requires { N > 0; }
 {
-  if(r.exp % 2 == 0)
-    return std::array{r.num, r.den, r.exp}; // already even (incl zero)
-
-  // safely make exp even, so it can be divided by 2 for square root
-  if(r.exp > 0)
-    return std::array{r.num * 10, r.den, r.exp - 1};
-  else
-    return std::array{r.num, r.den * 10, r.exp + 1};
+  if constexpr (N == 1) {
+    return r;
+  } else {
+    if (r.num == 0) {
+      return ratio(0);
+    }
+
+    const auto aligned = make_exp_align(r, N);
+    return ratio(iroot<N>(aligned[0]), iroot<N>(aligned[1]), aligned[2] / N);
+  }
 }
 
 }  // namespace detail
 
-[[nodiscard]] constexpr ratio sqrt(const ratio& r)
+template<std::intmax_t Num, std::intmax_t Den = 1>
+[[nodiscard]] constexpr ratio pow(const ratio& r) requires requires { Den != 0; }
```

Comment:
`Den != 0` to requires clause or as a named concept. Please prefer requires clause before the function signature.